### DISCUSSION TITLE
No unlock after canceling deployment.

### DISCUSSION THREAD

**Discussion by zimdin12** [deployphp/deployer#3336] [General]
I am getting error :
Executing task deploy:lock
In lock.php line 19:
                                                           
  Deploy locked.                                           
  Execute "dep deploy:unlock sitename.com" to unlock.
  
This started happening after I canceled the deployment. (gitlab job) 
I have multiple sites that are using this deployment script.
only the one (site) that got canceled started throwing this error.
I guess my only option is to make deployment with unlock as a first task ?

**Comment by Schrank** [comment]
> I guess my only option is to make deployment with unlock as a first task ?
 
This solves your problem, but allows running two deployments in parallel. So the alternative is to just remove `lock` and `unlock`.

I don't have a proper solution, because I don't know gitlab, but I'll try it anyway :-)

Technically the lock is a file created on the remote server, therefore if you kill the job the lock is not deleted.

Solution 1: Don't kill the pipeline but have a job which kills the pipeline (and therefore runs `unlock` before ending)
Solution 2: Have a `onAbort`(does this exist for gitlab?) or a pipeline which runs ALWAYS, even if a step returns something > 0 which runs only an unlock, so you can be sure, the unlock still happens AFTER the deployment failed/finished

**Comment by Schrank** [comment]
Maybe this? https://docs.gitlab.com/ee/ci/yaml/index.html#after_script

**Comment by zimdin12** [comment]
Since it doesn't happen often then I didn't modify my script (It shouldn't happen often)
But to get out of bad situation, where I cannot deploy, I did remove lock file from .dep folder (if I remember files and folders correctly).

**Reply by zimdin12** [reply]
To clarify, It only happened when I manually canceled a job, So I can go and remove lock file manually also.
I almost never cancel job, so jeah. But if somebody do know about "onAbort" in gitlab then please leave a comment :)

**Comment by Schrank** [comment]
I understand your Problem. I would love to have this myself.

Did you try to:

https://www.php.net/manual/en/function.register-shutdown-function.php

```php
register_shutdown_function(function() {
    invoke("deploy:failed");
});
```

Or something alike - untested :-)
